sdk/go/common/workspace/plugins.go

```diff
@@ -795,7 +796,109 @@ func (info PluginInfo) installLock() (unlock func(), err error) {
 
 // Install installs a plugin's tarball into the cache. See InstallWithContext for details.
 func (info PluginInfo) Install(tgz io.ReadCloser, reinstall bool) error {
-	return info.InstallWithContext(context.Background(), tgz, reinstall)
+	return info.InstallWithContext(context.Background(), tarPlugin{tgz}, reinstall)
+}
+
+type PluginContent interface {
+	io.Closer
+
+	WriteToDir(info PluginInfo) error
```

Comment:
I'd expect WriteToDir to take the directory path, not a PluginInfo. It also looks like things below are inconsistent about how they resolve from a PluginInfo to a directory.

Reply:
I see because your not assuming the file is already correctly named and so you rename the file to match the expected pulumi-kind-name format. I wonder if that information should be part of creating the FileSource instead? You have a PluginInfo on hand when calling `getFileSource` and so could pass the info and file into SingleFileSource with it just saving a copy of the kind and name.

Then WriteToDir could really just take a directory, rather than each source having to call info.PluginDir().

Reply:
That works. Done in https://github.com/pulumi/pulumi/pull/10094/commits/b1b61b842180008f33e5a1e9e154357d69d7fd8d.